""" CLI interface for brute force variant peptide caller """
import sys
import argparse
from typing import Set
from pathlib import Path
from moPepGen import seqvar, params
from moPepGen.cli import common as cli_common
from moPepGen.seqvar.VariantRecordPool import VariantRecordPool
from .caller import BruteForceVariantPeptideCaller
from .utils import create_mnvs, fix_indel_after_start_codon


# pylint: disable=W0212
def parse_args(subparsers:argparse._SubParsersAction):
    """ parse command line arguments """
    parser:argparse.ArgumentParser = subparsers.add_parser(
        name='bruteForce',
        help='Call variant peptide with the brute force algorithm.'
    )
    parser.add_argument(
        '-i', '--input-gvf',
        type=Path,
        help='GVF file',
        nargs="+"
    )
    parser.add_argument(
        '-r', '--reference-dir',
        type=Path,
        help='Reference directory. Must contain genome.fa, annotation.gtf, and'
        ' proteome.fasta. The directory should be generated by'
        ' the downsampleReference command'
    )
    valid_codon_tables = cli_common.get_valid_codon_tables()
    parser.add_argument(
        '--codon-table',
        type=str,
        default='Standard',
        choices=valid_codon_tables,
        help=f'Codon table. Defaults to "Standard". Supported codon tables: {valid_codon_tables}'
    )
    parser.add_argument(
        '--chr-codon-table',
        type=str,
        nargs='*',
        default=[],
        help='Chromosome specific codon table. Must be specified in the format of'
        ' "chrM:SGC1", where "chrM" is the chromosome name and "SGC1" is the codon'
        f' table to use to translate genes on chrM. Supported codon tables: {valid_codon_tables}.'
        ' By default, "SGC1" is assigned to mitochondrial chromosomes.'
    )
    parser.add_argument(
        '--start-codons',
        type=str,
        nargs='*',
        default=['ATG'],
        help='Default start codon(s) to use for novel ORF translation. Defaults to ["ATG"].'
    )
    parser.add_argument(
        '--chr-start-codons',
        type=str,
        nargs='*',
        default=[],
        help='Chromosome specific start codon(s). For example, "chrM:ATG,ATA,ATT".'
        'By defualt, mitochondrial chromosome name is automatically inferred and'
        'start codon "ATG", "ATA", "ATT", "ATC" and "GTG" are assigned to it.'
    )

    parser.add_argument(
        '-f', '--force',
        action='store_true',
        help='If not set, the program stops when there are more than 10'
        ' variants. When this flag is set, the program runs anyway. Noted that '
        ' the runtime is going to increase quickly after 10 variants.',
        default=False
    )
    parser.add_argument(
        '--variant-ids',
        type=str,
        help='List of variant labels.',
        nargs='*'
    )
    parser.add_argument(
        '--max-adjacent-as-mnv',
        type=int,
        help='Max number of adjacent SNPs or INDELs to be merged as a MNV.',
        default=2
    )
    parser.add_argument(
        '--selenocysteine-termination',
        action='store_true',
        help='Include peptides of selenoprotiens that the UGA is treated as '
        'termination instead of Sec.'
    )
    parser.add_argument(
        '--w2f-reassignment',
        action='store_true',
        help='Include peptides with W > F (Tryptophan to Phenylalanine) '
        'reassignment.'
    )
    cli_common.add_args_cleavage(parser)
    cli_common.add_args_debug_level(parser)

    parser.set_defaults(func=main)
    cli_common.print_help_if_missing_args(parser)
    return parser


def main(args):
    """ main """
    # Load genomic references
    ref_data = cli_common.load_references(
        args=argparse.Namespace(
            index_dir=None,
            genome_fasta=args.reference_dir/'genome.fasta',
            annotation_gtf=args.reference_dir/'annotation.gtf',
            proteome_fasta=args.reference_dir/'proteome.fasta',
            reference_source=None,
            codon_table=args.codon_table,
            chr_codon_table=args.chr_codon_table,
            start_codons=args.start_codons,
            chr_start_codons=args.chr_start_codons
        ),
        load_genome=True,
        load_canonical_peptides=False,
        load_codon_tables=True,
        load_proteome=True,
        invalid_protein_as_noncoding=True,
        check_protein_coding=True
    )

    # load GVF files
    variant_pool = seqvar.VariantRecordPool()
    variant_pool.anno = ref_data.anno
    for path in args.input_gvf:
        with open(path) as handle:
            variant_pool.load_variants(
                handle=handle,
                anno=ref_data.anno,
                genome=ref_data.genome
            )
    variant_pool = fix_indel_after_start_codon(variant_pool, ref_data)
    variant_pool = create_mnvs(variant_pool, args.max_adjacent_as_mnv)
    variant_peptides:Set[str] = set()
    for tx_id in variant_pool.data.keys():
        caller = BruteForceVariantPeptideCaller()
        caller.variant_pool = variant_pool
        caller.reference_data = ref_data
        caller.tx_id = tx_id
        caller.tx_model = caller.reference_data.anno.transcripts[caller.tx_id]
        caller.tx_seq = caller.tx_model.get_transcript_sequence(
            caller.reference_data.genome[caller.tx_model.transcript.chrom]
        )
        caller.w2f = args.w2f_reassignment
        caller.selenocysteine_termination = args.selenocysteine_termination

        caller.cleavage_params = params.CleavageParams(
            enzyme=args.cleavage_rule,
            exception=args.cleavage_exception,
            miscleavage=int(args.miscleavage),
            min_mw=float(args.min_mw),
            min_length=args.min_length,
            max_length=args.max_length
        )

        caller.create_canonical_peptide_pool()
        caller.get_start_index()

        caller.call_peptides()

        variant_peptides.update(caller.variant_peptides)

    for peptide in sorted(variant_peptides):
        print(peptide, file=sys.stdout)
